npm/pkg/dataplane/ipsets/ipset.go

```diff
@@ -0,0 +1,191 @@
+package ipsets
+
+import (
+	"fmt"
+
+	"github.com/Azure/azure-container-networking/npm/util"
+)
+
+type IPSet struct {
+	Name       string
+	HashedName string
+	Type       SetType
+	// IpPodKey is used for setMaps to store Ips and ports as keys
+	// and podKey as value
+	IPPodKey map[string]string
+	// This is used for listMaps to store child IP Sets
+	MemberIPSets map[string]*IPSet
+	// Using a map here to emulate set of netpol names
+	SelectorReference map[string]struct{}
+	NetPolReference   map[string]struct{}
+	IpsetReferCount   int
+}
+
+type SetType int32
+
+const (
+	Unknown                  SetType = 0
+	NameSpace                SetType = 1
+	KeyLabelOfNameSpace      SetType = 2
+	KeyValueLabelOfNameSpace SetType = 3
+	KeyLabelOfPod            SetType = 4
+	KeyValueLabelOfPod       SetType = 5
+	NamedPorts               SetType = 6
+	NestedLabelOfPod         SetType = 7
+	CIDRBlocks               SetType = 8
+)
+
+var SetTypeName = map[int32]string{
+	0: "Unknown",
+	1: "NameSpace",
+	2: "KeyLabelOfNameSpace",
+	3: "KeyValueLabelOfNameSpace",
+	4: "KeyLabelOfPod",
+	5: "KeyValueLabelOfPod",
+	6: "NamedPorts",
+	7: "NestedLabelOfPod",
+	8: "CIDRBlocks",
+}
+
+var SetTypeValue = map[string]int32{
+	"Unknown":                  0,
+	"NameSpace":                1,
+	"KeyLabelOfNameSpace":      2,
+	"KeyValueLabelOfNameSpace": 3,
+	"KeyLabelOfPod":            4,
+	"KeyValueLabelOfPod":       5,
+	"NamedPorts":               6,
+	"NestedLabelOfPod":         7,
+	"CIDRBlocks":               8,
+}
+
+func (x SetType) String() string {
+	return SetTypeName[int32(x)]
+}
+
+func GetSetType(x string) SetType {
+	return SetType(SetTypeValue[x])
+}
+
+type SetKind string
+
+const (
+	ListSet SetKind = "list"
+	HashSet SetKind = "set"
+)
+
+func NewIPSet(name string, setType SetType) *IPSet {
+	set := &IPSet{
+		Name:              name,
+		HashedName:        util.GetHashedName(name),
+		Type:              setType,
+		SelectorReference: make(map[string]struct{}),
+		NetPolReference:   make(map[string]struct{}),
+		IpsetReferCount:   0,
+	}
+	if getSetKind(set) == HashSet {
+		set.IPPodKey = make(map[string]string)
+	} else {
+		set.MemberIPSets = make(map[string]*IPSet)
+	}
+	return set
+}
+
+func (set *IPSet) GetSetContents() ([]string, error) {
+	contents := make([]string, 0)
+	setType := getSetKind(set)
+	switch setType {
+	case HashSet:
+		for podIP := range set.IPPodKey {
```

Comment:
also should allocate slice here with make([]string, len(set.MemberIPSets)), same below to avoid slice resizing, heh @rbtr

Reply:
Interesting, golint now complains "append to slice `contents` with non-zero initialized length (makezero)"

Reply:
yes, it's 👎🏼 to append to a slice that you've sized already.
to be maximally efficient _and_ satisfy the linter, you should add an idx and put the key values in to their spot directly:
```go
i := 0
keys := make([]string, len(set.IPPodKey)
for podIP := range set.IPPodKey {
	keys[i] = podIP
	i++
}
```

Reply:
for some reason this makes me uncomfortable, but yeah i see the perf improvement.

Reply:
Now i am thinking if golint will be okay with below 
> make([]string, 0, len(set.IPPodKey)

and then appending